app/controllers/foreman_ansible/concerns/api_common.rb

```diff
@@ -5,13 +5,44 @@ module Concerns
     # Shared task methods between api controllers
     module ApiCommon
       extend ActiveSupport::Concern
+      def update_associations
+        if controller_name == 'hosts'
+          join_model = HostAnsibleRole
+          foreign_key = :host_id
+          join_attrs = :host_ansible_roles_attributes
+        else
+          join_model = HostgroupAnsibleRole
+          foreign_key = :hostgroup_id
+          join_attrs = :hostgroup_ansible_roles_attributes
+        end
 
-      def find_ansible_roles
-        ids = params.fetch(:ansible_role_ids, []) || []
-        @ansible_roles = AnsibleRole.authorized(:view_ansible_roles).find(ids)
+        instance = instance_variable_get("@#{controller_name.singularize}")
+
+        attrs = find_ansible_roles_attributes params, join_model, foreign_key, instance
+        instance.update join_attrs => attrs
       rescue ActiveRecord::RecordNotFound => e
         not_found(e.message)
       end
+
+      def find_ansible_roles_attributes(params, join_model, foreign_key, instance)
```

Comment:
Wouldn't it be better to have this method on the host(group) as `ansible_role_ids=` I'd expect it to be the case.
As such setter make sense to me and given it takes Array, we can have the possition by idx as you have here, but I believe it could be easier if we don't need to transform it to nested parameters and you wouldn't need to transform params itself, what feels bit wrong to me.